### PR TITLE
ALSA: hda/hdmi: fix silent stream for first DP connection

### PR DESCRIPTION
**Pull request by kv2019i**
The silent streaming feature does not work as expected for the initial
playback case, when connection type is DisplayPort. Silent streaming
is activated, but until at least one real stream is played, no
converter is programmed and with DisplayPort, this results in audio
infoframes not to be sent out. Result is that initial samples of
audio playback are lost with receivers that require a preamble before
actual audio is played out.

Fix the issue by assigning a converter to the silent stream and
using existing i915_hsw_setup_stream() to configure the silent stream.

BugLink: https://github.com/thesofproject/linux/issues/2468
Fixes: 951894cf30f41 ("ALSA: hda/hdmi: Add Intel silent stream support")
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: sound/pci/hda/patch_hdmi.c [hunk 1]
```diff
@@ -1642,24 +1646,59 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+	 * TODO: we need to track cvt assignments for each pin separately
+	 *       and free the converter when silent streaming is ended
+	 */
+	per_cvt->assigned = 0;
+	per_pin->cvt_nid = per_cvt->cvt_nid;
+
+	codec_dbg(codec, "hdmi: using cvt %d (NID=0x%x) for pin NID=0x%x (dev=%d)\n",
+		  cvt_idx, per_cvt->cvt_nid, per_pin->pin_nid, per_pin->dev_id);
+
+	snd_hda_set_dev_select(codec, per_pin->pin_nid, per_pin->dev_id);
```
Comment (harshapriya-n) [inline sound/pci/hda/patch_hdmi.c:1689]: check the return value requried?

File: sound/pci/hda/patch_hdmi.c [hunk 2]
```diff
@@ -1642,24 +1646,59 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+	snd_hda_codec_write_cache(codec, per_pin->pin_nid, 0,
+				  AC_VERB_SET_CONNECT_SEL,
+				  per_pin->mux_idx);
+
+	/* configure unused pins to choose other converters */
+	pin_cvt_fixup(codec, per_pin, 0);
```
Comment (harshapriya-n) [inline sound/pci/hda/patch_hdmi.c:1695]: Does this function call into the intel_not_share_assigned_cvt() that calls the same sequence of hdmi_choose_cvt, get_cvt and snd_hda_set_dev_select?

**Comment by kv2019i** [comment]
V2 uploaded. This now works better in multi-monitor scenarios, but I'm still able to trigger some errors, so more work to be done. Sharing interim status.

File: sound/pci/hda/patch_hdmi.c (hunk 2 above)
Comment (kv2019i) [reply sound/pci/hda/patch_hdmi.c:1695]: @harshapriya-n It's a similar chain on calls, but pin_cvt_fixup does the calls for other pins. I.e. if we set up a pin+converter for silent streaming, we should ensure other pins are not connected to this same converter. If not done, a mute set on some other pin, could mute the silent stream and break the functionality.

File: sound/pci/hda/patch_hdmi.c (hunk 1 above)
Comment (kv2019i) [reply sound/pci/hda/patch_hdmi.c:1689]: It seems majority of the calls to set_dev_select are without error checking, so following the same convention here. Not at least for this patch to address.

**Comment by kv2019i** [comment]
V3 update to the patch. I now managed to fix a set of multi-monitor/MST problems. More testing is still needed, but this is starting to be more mature solution. Testing feedback is welcome.

File: sound/pci/hda/patch_hdmi.c [hunk 3]
```diff
@@ -1642,30 +1650,86 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+#define INTEL_SILENTSTR_CHANNELS		2
+#define INTEL_SILENTSTR_FORMAT		SNDRV_PCM_FORMAT_S16_LE
+#define INTEL_SILENTSTR_TAG		0xf
+
 static void silent_stream_enable(struct hda_codec *codec,
-				struct hdmi_spec_per_pin *per_pin)
+				 struct hdmi_spec_per_pin *per_pin)
 {
-	unsigned int newval, oldval;
+	struct hdmi_spec_per_cvt *per_cvt = NULL;
```
Comment (lyakh) [inline sound/pci/hda/patch_hdmi.c]: superfluous initialisation

File: sound/pci/hda/patch_hdmi.c [hunk 4]
```diff
@@ -1642,30 +1650,86 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+	snd_hda_codec_setup_stream(codec, per_pin->cvt_nid, INTEL_SILENTSTR_TAG, 0, format);
+
+ unlock_out:
 	mutex_unlock(&per_pin->lock);
 }
 
+static void silent_stream_disable(struct hda_codec *codec,
+				  struct hdmi_spec_per_pin *per_pin)
+{
+	struct hdmi_spec_per_cvt *per_cvt = NULL;
```
Comment (lyakh) [inline sound/pci/hda/patch_hdmi.c]: ditto

File: sound/pci/hda/patch_hdmi.c [hunk 5]
```diff
@@ -1642,30 +1650,86 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+	int cvt_idx;
+
+	if (!per_pin->silent_stream)
+		return;
+
+	codec_dbg(codec, "HDMI: disable silent stream on cvt-NID=0x%x\n", per_pin->cvt_nid);
+	cvt_idx = cvt_nid_to_cvt_index(codec, per_pin->cvt_nid);
+	per_cvt = get_cvt(spec, cvt_idx);
+	WARN_ON(!per_cvt);
+	if (per_cvt)
```
Comment (lyakh) [inline sound/pci/hda/patch_hdmi.c]: could do `if (!WARN_ON(!per_cvt))` but maybe that'd be a bit too many negations

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/pci/hda/patch_hdmi.c (hunk 3 above)
Comment (kv2019i) [reply sound/pci/hda/patch_hdmi.c]: Thanks @lyakh will fix.

File: sound/pci/hda/patch_hdmi.c (hunk 5 above)
Comment (kv2019i) [reply sound/pci/hda/patch_hdmi.c]: True. Also get_cvt() doesn't return a null, so I'll rework this in any case.

File: sound/pci/hda/patch_hdmi.c [hunk 6]
```diff
@@ -1642,30 +1650,93 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
...
+	int cvt_idx, pin_idx, err;
+	unsigned int format;
 
 	mutex_lock(&per_pin->lock);
 
 	if (!per_pin->channels)
-		per_pin->channels = 2;
+		per_pin->channels = I915_SILENT_CHANNELS;
+
+	if (per_pin->setup) {
```
Comment (kv2019i) [inline sound/pci/hda/patch_hdmi.c:1669]: Move this check above L1669

**Comment by kv2019i** [comment]
Sent to upstream now, review in progress.

**Comment by kv2019i** [comment]
Merged upstream, closing.

**Pull request closed without merging**